src/pykeen/datasets/analysis.py

```diff
@@ -0,0 +1,449 @@
+    # for each group, yield from skyline
+    for (r_id, pat), values in data.items():
+        yield from (
+            (r_id, pat, supp, conf)
+            for supp, conf in _get_skyline(values)
+        )
+
+
+def _composition_candidates(
+    mapped_triples_list: Collection[Tuple[int, int, int]],
+) -> Collection[Tuple[int, int]]:
+    r"""
+    Pre-filtering relation pair candidates for composition pattern.
+
+    Determines all relation pairs (r, r') with at least one entity e such that
+
+    .. math ::
+        \{(h, r, e), (e, r', t)\} \subseteq \mathcal{T}
+
+    :param mapped_triples_list:
+        The collection of ID-based triples.
+
+    :return:
+        A set of relation pairs.
+    """
+    # index triples
+    # incoming relations per entity
+    ins: Mapping[int, Set[int]] = defaultdict(set)
+    # outgoing relations per entity
+    outs: Mapping[int, Set[int]] = defaultdict(set)
+    for h, r, t in mapped_triples_list:
+        outs[h].add(r)
+        ins[t].add(r)
+
+    # return candidates
+    return {
+        (r1, r2)
+        for e, r1s in ins.items()
+        for r1 in r1s
+        for r2 in outs[e]
+    }
+
+
+def _yield_unary_patterns(
+    pairs: Mapping[int, Set[Tuple[int, int]]],
+) -> Iterable[Tuple[int, str, int, float]]:
+    r"""
+    Yield unary patterns from pre-indexed triples.
+
+    Patterns:
+        Symmetry
+            .. math ::
+                r(x, y) \implies r(y, x)
+
+        Anti-Symmetry
+            .. math ::
+                r(x, y) \implies \neg r(y, x)
+
+    :param pairs:
+        A mapping from relations to the set of entity pairs.
+
+    :yields:
+        A tuple (relation_id, pattern_type, support, confidence).
+    """
+    for r, ht in pairs.items():
+        support = len(ht)
+        rev_ht = {(t, h) for h, t in ht}
+        confidence = len(ht.intersection(rev_ht)) / support
+        yield r, "symmetry", support, confidence
+        confidence = len(ht.difference(rev_ht)) / support
+        yield r, "anti-symmetry", support, 1 - confidence
+
+
+def _yield_binary_patterns(
+    pairs: Mapping[int, Set[Tuple[int, int]]],
+) -> Iterable[Tuple[int, str, int, float]]:
+    r"""
+    Yield binary patterns from pre-indexed triples.
+
+    Patterns:
+        Inversion
+        .. math ::
+            r'(x, y) \implies r(y, x)
+
+    :param pairs:
+        A mapping from relations to the set of entity pairs.
+
+    :yields:
+        A tuple (relation_id, pattern_type, support, confidence).
+    """
+    for (r1, ht1), (r, ht2) in itt.combinations(pairs.items(), r=2):
+        support = len(ht1)
+        confidence = len(ht1.intersection(ht2)) / support
+        yield r, "inversion", support, confidence
+
+
+def _yield_ternary_patterns(
+    mapped_triples_list: Collection[Tuple[int, int, int]],
+    pairs: Mapping[int, Set[Tuple[int, int]]],
+) -> Iterable[Tuple[int, str, int, float]]:
+    r"""
+    Yield ternary patterns from pre-indexed triples.
+
+    Patterns:
+        Composition
```

Comment:
can we also do transitivity? 

`r(x, y) \land r(y, z) \implies r(x, z)`

Reply:
or also anti-transitivity (I made this word up)

`r(x, y) \land r(y, z) \implies r'(x, z)`

Example: X decreases Y. Y decreases Z. Imply X increases Z.

Reply:
I guess so :sweat_smile: For now, I added only the ones from the RotatE paper, also requested by one of the reviewers of the benchmarking paper.

However, the "data structure" (relation_id, pattern, support, confidence) allows relatively easy extension to other patterns.